pkg/kubelet/kubelet.go

```diff
@@ -2790,76 +2787,42 @@ func (kl *Kubelet) canResizePod(pod *v1.Pod) (bool, *v1.Pod, v1.PodResizeStatus)
 		}
 	}
 
-	if ok, failReason, failMessage := kl.canAdmitPod(otherActivePods, podCopy); !ok {
+	if ok, failReason, failMessage := kl.canAdmitPod(otherActivePods, pod); !ok {
 		// Log reason and return. Let the next sync iteration retry the resize
-		klog.V(3).InfoS("Resize cannot be accommodated", "pod", podCopy.Name, "reason", failReason, "message", failMessage)
-		return false, podCopy, v1.PodResizeStatusDeferred
+		klog.V(3).InfoS("Resize cannot be accommodated", "pod", klog.KObj(pod), "reason", failReason, "message", failMessage)
+		return false, v1.PodResizeStatusDeferred
 	}
 
-	for _, container := range podCopy.Spec.Containers {
-		idx, found := podutil.GetIndexOfContainerStatus(podCopy.Status.ContainerStatuses, container.Name)
-		if found {
-			for rName, rQuantity := range container.Resources.Requests {
-				podCopy.Status.ContainerStatuses[idx].AllocatedResources[rName] = rQuantity
-			}
-		}
-	}
-	return true, podCopy, v1.PodResizeStatusInProgress
+	return true, v1.PodResizeStatusInProgress
 }
 
-func (kl *Kubelet) handlePodResourcesResize(pod *v1.Pod) *v1.Pod {
-	if pod.Status.Phase != v1.PodRunning {
-		return pod
-	}
-	podResized := false
-	for _, container := range pod.Spec.Containers {
-		if len(container.Resources.Requests) == 0 {
-			continue
-		}
-		containerStatus, found := podutil.GetContainerStatus(pod.Status.ContainerStatuses, container.Name)
-		if !found {
-			klog.V(5).InfoS("ContainerStatus not found", "pod", pod.Name, "container", container.Name)
-			break
-		}
-		if len(containerStatus.AllocatedResources) != len(container.Resources.Requests) {
-			klog.V(5).InfoS("ContainerStatus.AllocatedResources length mismatch", "pod", pod.Name, "container", container.Name)
-			break
-		}
-		if !cmp.Equal(container.Resources.Requests, containerStatus.AllocatedResources) {
-			podResized = true
-			break
-		}
-	}
-	if !podResized {
-		return pod
+func (kl *Kubelet) handlePodResourcesResize(pod *v1.Pod) (*v1.Pod, error) {
+	allocatedPod, updated := kl.statusManager.UpdatePodFromAllocation(pod)
+	if !updated {
+		// Pod is not resizing, nothing more to do here.
+		return pod, nil
 	}
 
 	kl.podResizeMutex.Lock()
 	defer kl.podResizeMutex.Unlock()
-	fit, updatedPod, resizeStatus := kl.canResizePod(pod)
-	if updatedPod == nil {
-		return pod
-	}
+	fit, resizeStatus := kl.canResizePod(pod)
 	if fit {
 		// Update pod resource allocation checkpoint
-		if err := kl.statusManager.SetPodAllocation(updatedPod); err != nil {
-			//TODO(vinaykul,InPlacePodVerticalScaling): Can we recover from this in some way? Investigate
-			klog.ErrorS(err, "SetPodAllocation failed", "pod", klog.KObj(updatedPod))
-			return pod
+		if err := kl.statusManager.SetPodAllocation(pod); err != nil {
+			return nil, err
 		}
+	} else {
+		// If resize isn't immediately feasible, proceed with the allocated pod.
```

Comment:
if resizing is feasible, the allocated pod would be the same as the pod, no? Why separate the two?

Reply:
I'm not sure I totally understand what you're suggesting, but I flipped this around to hopefully make it more clear. If the pod doesn't fit, we should proceed with the allocated pod. If it does fit, then we set the allocation to be the new state. Is that what you meant?

Reply:
My mental model (which might be wrong):

1. If resizing is feasible, update the allocated resources of the pod
2. If resizing is not feasible, the allocated resources stays the same

Either way, we'd be looking at "(updated) allocated resources" of the pod. I think the naming of the variables here might be my source of confusion....


Another question I have is that do we always set "allocated resources" even when there's no resize request? If not, doing that might help simplify some logic in the code.

Reply:
Yeah. I updated more references to be to the `allocatedPod`, hopefully it's more clear now.

> Another question I have is that do we always set "allocated resources" even when there's no resize request? If not, doing that might help simplify some logic in the code.

The first part of the function gets the allocated pod, but if there's no resize (no update), it just returns early with that. Is that what you mean?

Would it help to rename this function? Maybe `handlePodAllocation` instead?

Reply:
> The first part of the function gets the allocated pod, but if there's no resize (no update), it just returns early with that. Is that what you mean?

My bad. That was a more general question. When kubelet sees a pod for the first time with the desired spec (and no resize request yet), do we also set the allocated resources in the status? 

(BTW, this question doesn't need to block the PR)

Reply:
The allocated resources are just the admitted resources. So if a new pod is admitted but doesn't have a resize request, then we still set its allocated resources even though they match the spec. Same goes for the container status (although there are a couple changes to the way that API is handled, which I'm tackling in a follow-up PR)